README.md
# desafio01

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2021 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2021-11-22
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2021-11-22
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/desafio01/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/desafio01/compare/0.1.0...0.1.1


test/desafio01/core_test.clj
(ns desafio01.core-test
  (:require [clojure.test :refer :all]
            [desafio01.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


test/desafio01/desafio02_test.clj
(ns desafio01.desafio02-test
  (:require [clojure.test :refer :all]
            [desafio01.desafio02 :refer :all]
            [desafio01.model :as d.model]
            [schema.core :as s]))

(def cliente {:nome "Maria da Silva" :cpf "12356743565" :email "maria@gmal.com"
                :cartao {:numero  "1324567897643256" :cvv "123" :validade "03/2030" :limite 2000.00
                             :compras [{:data "01/11/2021" :valor 200.00 :estabelecimento "Extra" :categoria "Alimentação"}
                                       {:data "23/09/2020" :valor 545.00 :estabelecimento "Hospital Sao luiz" :categoria "Saúde"}
                                       {:data "03/04/2021" :valor 78.67 :estabelecimento "Extra" :categoria "Alimentação"}
                                       {:data "15/01/2020" :valor 21.30 :estabelecimento "Drograsil" :categoria "Saúde"}
                                       {:data "12/11/2021" :valor 567.99 :estabelecimento "Wise up" :categoria "Educação"}]}})

(def cliente-sem-compras {:nome "Maria da Silva" :cpf "12356743565" :email "maria@gmal.com"
                          :cartao {:numero  "1324567897643256" :cvv "123" :validade "03/2030" :limite 2000.00
                                   :compras []}})

(def cliente-com-apenas-uma-compra {:nome "Maria da Silva" :cpf "12356743565" :email "maria@gmal.com"
              :cartao {:numero  "1324567897643256" :cvv "123" :validade "03/2030" :limite 2000.00
                       :compras [{:data "01/11/2021" :valor 200.00 :estabelecimento "Extra" :categoria "Alimentação"}]}})

(def cliente-com-nova-compra {:nome "Maria da Silva" :cpf "12356743565" :email "maria@gmal.com"
                                    :cartao {:numero  "1324567897643256" :cvv "123" :validade "03/2030" :limite 2000.00
                                             :compras [{:data "12/11/2021" :valor 333.99 :estabelecimento "Ricoy" :categoria "Alimentacao"}]}})

(def nova-compra {:data "12/11/2021" :valor 333.99 :estabelecimento "Ricoy" :categoria "Alimentacao"})


(def compras-acima-de-duzentos [{:data "01/11/2021" :valor 200.00 :estabelecimento "Extra" :categoria "Alimentação"}
                                {:data "23/09/2020" :valor 545.00 :estabelecimento "Hospital Sao luiz" :categoria "Saúde"}
                                {:data "12/11/2021" :valor 567.99 :estabelecimento "Wise up" :categoria "Educação"}] )

(def compras-no-extra [{:data "01/11/2021" :valor 200.00 :estabelecimento "Extra" :categoria "Alimentação"}
                       {:data "03/04/2021" :valor 78.67 :estabelecimento "Extra" :categoria "Alimentação"}])

(def totais-por-categoria [{:categoria "Alimentação", :valor 278.67 }
                           {:categoria "Saúde", :valor 566.30 }
                           {:categoria "Educação", :valor 567.99 }])

(def compra-com-valor-nulo {:data "01/11/2021" :valor nil :estabelecimento "Extra" :categoria "Alimentação"})



;TESTES PARA  LISTAR COMPRAS

(deftest busca-compras-test

  (testing "Listar compras maior ou igual ao valor"
   (is (= compras-acima-de-duzentos
          (listar-compras cliente, 200))))

  (testing "Retornar vazio para valor inexistente"
    (is (= []
           (listar-compras cliente,600))))

  (testing "Listar compras por estabelecimento"
    (is (= compras-no-extra
           (listar-compras cliente, "Extra"))))

  (testing "Retornar vazio para estabelecimento inexistente"
    (is (= []
           (listar-compras cliente,"Pao de acucar"))))

  (testing "Retornar vazio para cliente sem compras"
    (is (=  []
           (listar-compras cliente-sem-compras, 0)))))

;TESTES GASTOS POR CATEGORIA
  (deftest totais-por-categoria-test

    (testing "Retorna totais por categoria"
      (is (= totais-por-categoria
             (calcula-totais-por-categoria cliente))))

    (testing "Retorna total para cliente com apenas uma compra"
      (is (= [{:categoria "Alimentação", :valor 200.00}]
             (calcula-totais-por-categoria cliente-com-apenas-uma-compra))))

    (testing "Retorna vazio caso nao exista compras"
      (is (= []
             (calcula-totais-por-categoria cliente-sem-compras)))))

;TESTES ADICAO DE NOVA COMPRA
 (deftest adiciona-compra-test

   (testing "Adicao de nova compra"
     (is (= cliente-com-nova-compra
            (adiciona-compra cliente-sem-compras nova-compra))))

   (testing "Nao adiciona compra ja existente"
     (is (= cliente-com-nova-compra
            (adiciona-compra cliente-com-nova-compra nova-compra))))

   (testing "Lanca excecao ao receber compra com valor nulo"
     (is (thrown? clojure.lang.ExceptionInfo
           (adiciona-compra cliente-sem-compras compra-com-valor-nulo))))

   (testing "Lanca excecao ao receber compra com valor negativo"
     (is (thrown? clojure.lang.ExceptionInfo
                  (adiciona-compra cliente-sem-compras compra-com-valor-nulo))))
   )


src/desafio01/model.clj
(ns desafio01.model
  (:require [schema.core :as s]))

(s/def PosNum (s/pred #(or (pos? %) (zero? %))))


(s/def Compra {:data s/Str, :valor PosNum, :estabelecimento s/Str, :categoria s/Str })

(s/def Cartao {:numero s/Str, :cvv s/Str, :validade s/Str, :limite s/Num, :compras [Compra]})

(s/def Cliente {:nome s/Str, :cpf s/Str, :email s/Str, :cartao Cartao})

(s/def ResumoCategoria {:categoria s/Str, :valor s/Num})

;(println (s/validate Compra
;                     {:data            "01/11/2021"
;                      :valor            200.00
;                      :estabelecimento "Extra super SA"
;                      :categoria       "Alimentação"}))

;(println (s/validate Cartao {:numero  "1324567897643256" :cvv "123" :validade "03/2030" :limite 2000.00
;                             :compras [{:data "01/11/2021" :valor 200.00 :estabelecimento "Extra super SA" :categoria "Alimentação"}
;                                       {:data "23/09/2020" :valor 545.00 :estabelecimento "Extra super SA" :categoria "Saúde"}
;                                       {:data "03/04/2021" :valor 78.67 :estabelecimento "Extra super SA" :categoria "Alimentação"}
;                                       {:data "15/01/2020" :valor 21.30 :estabelecimento "Extra super SA" :categoria "Saúde"}
;                                       {:data "12/11/2021" :valor 567.99 :estabelecimento "Extra super SA" :categoria "Educação"}]}))
;
;(println (s/validate Cliente { :nome "Maria da Silva" :cpf "12356743565" :email "maria@gmal.com"
;                             :cartao     {:numero  "1324567897643256" :cvv "123" :validade "03/2030" :limite 2000.00
;                                          :compras [{:data "01/11/2021" :valor 200.00 :estabelecimento "Extra super SA" :categoria "Alimentação"}
;                                                    {:data "23/09/2020" :valor 545.00 :estabelecimento "Extra super SA" :categoria "Saúde"}
;                                                    {:data "03/04/2021" :valor 78.67 :estabelecimento "Extra super SA" :categoria "Alimentação"}
;                                                    {:data "15/01/2020" :valor 21.30 :estabelecimento "Extra super SA" :categoria "Saúde"}
;                                                    {:data "12/11/2021" :valor 567.99 :estabelecimento "Extra super SA" :categoria "Educação"}]}}))

src/desafio01/desafio01.clj
(ns desafio01.desafio01
  (:use clojure.pprint)
  (:require [java-time :as time]))

(def cliente01 {:nome "Maria da Silva" :cpf "12356743565" :email "maria@gmal.com"
                :cartao     {:numero  "1324567897643256" :cvv "123" :validade "03/2030" :limite 2000.00
                             :compras [{:data "01/11/2021" :valor 200.00 :estabelecimento "Extra super SA" :categoria "Alimentação"}
                                       {:data "23/09/2020" :valor 545.00 :estabelecimento "Extra super SA" :categoria "Saúde"}
                                       {:data "03/04/2021" :valor 78.67 :estabelecimento "Extra super SA" :categoria "Alimentação"}
                                       {:data "15/01/2020" :valor 21.30 :estabelecimento "Extra super SA" :categoria "Saúde"}
                                       {:data "12/11/2021" :valor 567.99 :estabelecimento "Extra super SA" :categoria "Educação"}]}})

;OBTER COMPRAS

(defn obter-compras [cliente]
  (-> cliente
      :cartao
      :compras))


; AGRUPAMENTO POR CATEGORIA
(defn calcula-total-da-categoria [[categoria compras]]
  (let [valor-total-da-categoria (reduce + (map :valor compras))]
    [categoria valor-total-da-categoria]))

(defn calcula-gastos-totais-por-categoria
  [compras]
  (map calcula-total-da-categoria compras))

(defn agrupa-por-categoria [cliente]
  (->> (obter-compras cliente)
       (group-by :categoria)
       (calcula-gastos-totais-por-categoria)
       (pprint)))

;(agrupa-por-categoria cliente01)

;FATURA DO MES
(defn mes-desejado? [compra periodo]
  (let [mes-corrente periodo
        mes-da-fatura (time/format "MM/yyyy" (time/local-date "dd/MM/yyyy" (:data compra)))]
    (= mes-corrente mes-da-fatura))
  )

(defn calcula-total-do-mes [compras periodo]
  (reduce + (map :valor (filter #(mes-desejado? % periodo) compras))))

(defn calcula-valor-da-fatura [cliente periodo]
  (-> (obter-compras cliente)
      (calcula-total-do-mes periodo)
      (pprint)))

;(calcula-valor-da-fatura cliente01 "04/2021")

;BUSCA DE COMPRAS PELO ESTABELECIMENTO OU VALOR

(defn corresponde? [compra filtro]
  (let [estabelecimento (:estabelecimento compra)
        valor (:valor compra)]
    (if (string? filtro)
      (= estabelecimento filtro)
      (>= valor filtro))))

(defn aplica-filtro [compras filtro]
  (filter #(corresponde? % filtro) compras))

(defn busca-compras [cliente filtro]
  (-> (obter-compras cliente)
      (aplica-filtro filtro)
      (pprint))
  )

;(busca-compras cliente01 545M)












src/desafio01/desafio02.clj
(ns desafio01.desafio02
  (:use clojure.pprint)
  (:require [java-time :as time]
            [schema.core :as s]
            [desafio01.model :as d.model]))

(s/set-fn-validation! true)

;OBTER COMPRAS
(defn obter-compras [cliente]
  (-> cliente
      :cartao
      :compras))


; TOTAIS POR CATEGORIA
(s/defn calcula-total-por-categoria :- d.model/ResumoCategoria
  [[categoria compras]]
  (let [valor-total (reduce + (map :valor compras))]
    {:categoria categoria :valor valor-total}))

(s/defn calcula-totais-por-categoria :- [d.model/ResumoCategoria]
  [cliente :- d.model/Cliente]
  (let [compras (obter-compras cliente)
        compras-por-categoria (group-by :categoria compras)]
    (map calcula-total-por-categoria compras-por-categoria)))


;FATURA DO MES

(s/defn mes-desejado? :- s/Bool
  [compra :- d.model/Compra, periodo :- s/Str]
  (let [mes-corrente periodo
        mes-da-fatura (time/format "MM/yyyy" (time/local-date "dd/MM/yyyy" (:data compra)))]
    (= mes-corrente mes-da-fatura)))

(s/defn calcula-valor-da-fatura :- s/Num
  [cliente :- d.model/Cliente, periodo :- s/Str]
  (let [compras (obter-compras cliente)]
        (reduce + (map :valor (filter #(mes-desejado? % periodo) compras)))))

;LISTAR COMPRAS PELO ESTABELECIMENTO OU VALOR
(s/defn corresponde? :- s/Bool
  [compra :- d.model/Compra, filtro]
  (let [estabelecimento (:estabelecimento compra)
        valor (:valor compra)]
    (if (string? filtro)
      (= estabelecimento filtro)
      (>= valor filtro))))


(s/defn listar-compras :- [d.model/Compra]
  [cliente :- d.model/Cliente, filtro]
  (let [compras (obter-compras cliente)]
    (filter #(corresponde? % filtro) compras)))


;ADICIONAR COMPRA

(s/defn compra-repetida?
  [compra :- d.model/Compra, cliente :- d.model/Cliente]
  (some #(= compra %) (obter-compras cliente)))

(s/defn adiciona-compra :- d.model/Cliente
  [cliente :- d.model/Cliente, compra :- d.model/Compra]
  (if (compra-repetida? compra cliente)
    cliente
    (update-in cliente [:cartao :compras] conj compra)))




src/desafio01/principal.clj
(ns desafio01.principal
  (:use clojure.pprint)
  (:require [java-time :as time]))

(def cliente01 {:nome "Maria da Silva" :cpf "12356743565" :email "maria@gmal.com"
                :cartao     {:numero  "1324567897643256" :cvv "123" :validade "03/2030" :limite 2000.00
                             :compras [{:data "01/11/2021" :valor 200.00 :estabelecimento "Extra super SA" :categoria "Alimentação"}
                                       {:data "23/09/2020" :valor 545.00 :estabelecimento "Extra super SA" :categoria "Saúde"}
                                       {:data "03/04/2021" :valor 78.67 :estabelecimento "Extra super SA" :categoria "Alimentação"}
                                       {:data "15/01/2020" :valor 21.30 :estabelecimento "Extra super SA" :categoria "Saúde"}
                                       {:data "12/11/2021" :valor 567.99 :estabelecimento "Extra super SA" :categoria "Educação"}]}})

;OBTER COMPRAS

(defn obter-compras [cliente]
  (-> cliente
      :cartao
      :compras))


; AGRUPAMENTO POR CATEGORIA
(defn calcula-total-da-categoria [[categoria compras]]
  (let [valor-total-da-categoria (reduce + (map :valor compras))]
    [categoria valor-total-da-categoria]))

(defn calcula-gastos-totais-por-categoria
  [compras]
  (map calcula-total-da-categoria compras))

(defn agrupa-por-categoria [cliente]
  (->> (obter-compras cliente)
       (group-by :categoria)
       (calcula-gastos-totais-por-categoria)
       (pprint)))

;(agrupa-por-categoria cliente01)

;FATURA DO MES
(defn mes-desejado? [compra periodo]
  (let [mes-corrente periodo
        mes-da-fatura (time/format "MM/yyyy" (time/local-date "dd/MM/yyyy" (:data compra)))]
    (= mes-corrente mes-da-fatura))
  )

(defn calcula-total-do-mes [compras periodo]
  (reduce + (map :valor (filter #(mes-desejado? % periodo) compras))))

(defn calcula-valor-da-fatura [cliente periodo]
  (-> (obter-compras cliente)
      (calcula-total-do-mes periodo)
      (pprint)))

;(calcula-valor-da-fatura cliente01 "04/2021")

;BUSCA DE COMPRAS PELO ESTABELECIMENTO OU VALOR

(defn corresponde? [compra filtro]
  (let [estabelecimento (:estabelecimento compra)
        valor (:valor compra)]
    (if (string? filtro)
      (= estabelecimento filtro)
      (>= valor filtro))))

(defn aplica-filtro [compras filtro]
  (filter #(corresponde? % filtro) compras))

(defn busca-compras [cliente filtro]
  (-> (obter-compras cliente)
      (aplica-filtro filtro)
      (pprint))
  )

;(busca-compras cliente01 545M)










